### ISSUE TITLE
Home-App takes long to update

### ISSUE THREAD

**Issue by OnkelJudith** [andig/homebridge-fritz#30]
Hi,

after installing and configuring homebridge-fritz all my devices are found in my Apple Home app. But it takes a long time to update temperate and states (about 10-30 seconds). Everytime the app is minimized and reopened it's updating again.
Is this a normal behaviour? Does anyone else have these problems? My devices are working perfectly except this long update.

Heres my config.json

`{
    "bridge":{
        "name":"Homebridge",
        "username":"11:22:33:44:55:66",
        "port":51826,
        "pin":"012-34-567",
        "manufacturer": "@nfarina",
        "model": "Homebridge",
        "serialNumber": 0.4.20"
    },
    "description":"Homebridge Server.",
    "platforms":[
        {
            "platform":"Fritz!Box",
            "name":"7490",
            "username":"habridge",
            "password":"xxxxxxx",
            "url":"http://fritz.box",
            "interval":30,
            "options":{
                "strictSSL":false
            }
        }
    ]
}
`

**Comment by andig** [comment]
I can see two potential problems:

  - home app should not update all devices on minimize/reopen- could you check if its doing that for devices of other platforms, too?
  - updating should be faster- for your fritzbox you could check out this branch https://github.com/andig/homebridge-fritz/tree/automode and test if the new `concurent` setting for parallel updates is helping

**Comment by OnkelJudith** [comment]
Thanks for your feedback. Unfortunately I don‘t own other devices.
After updating to this branche the update process seems to be slightly faster. I guess there are some packages missing or misconfiguered on my raspberry.

**Comment by OnkelJudith** [comment]
After reinstallalation on my second raspberry (ans upgrading node from v4 to v8 everything works smoother. I think this issue can be closed. Can‘t name exactly the problem. Guessing some packages led to this.

**Comment by andig** [comment]
Good to hear. Lets keep this one open since I‘ve had one other report of this behaviour.

**Comment by jngmrks** [comment]
About 30 sec. every start and another 30 sec. for each update. Seems a way too long.
Updates running each time the app is loaded from background.

**Comment by xkonni** [comment]
having the same issues. the platform always updates when starting 'home' or the quick launch.

**Comment by andig** [comment]
@jngmrks @xkonni 

Could you please provide a homebridge logfile that shows the behaviour? I'm interested in the time when the app is loaded from background:

    DEBUG=* ./bin/homebridge

Also see @OnkelJudith's comment above- which node version are you running?

You can also try the latest development version which supports running multiple Fritz API requests in parallel:

     npm install --unsafe-perm git://github.com/andig/homebridge-fritz#automode

Then enable the `concurrent` setting in the platform configuration.

**Comment by jngmrks** [comment]
Node 7.9.0, logfile needs some time.

**Comment by xkonni** [comment]
Hey Andi, 

thanks for the quick reply!
running the latest homebridge, node v8.8.1 and i did update your plugin to the automode branch 
already. 
before posting i didn't yet enable concurrent, but that doesn't seem to change anything.

tried without fritz, to be sure, and that removes the delay.

attached a log as you suggested.
* normal
  [homebridge.log](https://github.com/andig/homebridge-fritz/files/1545674/homebridge.log)

* concurrent = true
  [homebridge_conc.log](https://github.com/andig/homebridge-fritz/files/1545676/homebridge_conc.log)

**Comment by jngmrks** [comment]
> tried without fritz, to be sure, and that removes the 
> delay.

xkonni, what do you mean with this?

**Comment by jngmrks** [comment]
````
[12/10/2017, 2:10:54 PM] Homebridge is running on port 51826.
-- home app starts --
[12/10/2017, 2:10:58 PM] [Fritz!Box] Getting thermostat 109711038560 target heating state
[12/10/2017, 2:10:58 PM] [Fritz!Box] Getting thermostat 109711038560 temperature
[12/10/2017, 2:10:58 PM] [Fritz!Box] Getting thermostat 109711038560 target temperature
[12/10/2017, 2:10:58 PM] [Fritz!Box] Getting thermostat 109711038560 heating state
[12/10/2017, 2:10:58 PM] [Fritz!Box] Getting thermostat 109711038560 battery status
[12/10/2017, 2:10:58 PM] [Fritz!Box] Getting guest WLAN state
[12/10/2017, 2:10:58 PM] [Fritz!Box] Getting thermostat 117950252408 target heating state
[12/10/2017, 2:10:58 PM] [Fritz!Box] Getting thermostat 117950252408 temperature
[12/10/2017, 2:10:58 PM] [Fritz!Box] Getting thermostat 117950252408 heating state
[12/10/2017, 2:10:58 PM] [Fritz!Box] Getting thermostat 117950252408 target temperature
[12/10/2017, 2:10:58 PM] [Fritz!Box] Getting thermostat 117950252408 battery status
[12/10/2017, 2:10:58 PM] [Fritz!Box] Getting thermostat 117950210296 target heating state
[12/10/2017, 2:10:58 PM] [Fritz!Box] Getting thermostat 117950210296 target temperature
[12/10/2017, 2:10:58 PM] [Fritz!Box] Getting thermostat 117950210296 temperature
[12/10/2017, 2:10:58 PM] [Fritz!Box] Getting thermostat 117950210296 heating state
[12/10/2017, 2:10:58 PM] [Fritz!Box] Getting thermostat 117950210296 battery status
[12/10/2017, 2:10:58 PM] [Fritz!Box] Getting outlet 116570017928 state
[12/10/2017, 2:10:58 PM] [Fritz!Box] Getting outlet 116570017928 temperature
[12/10/2017, 2:10:58 PM] [Fritz!Box] Getting thermostat 117950212336 target temperature
[12/10/2017, 2:10:58 PM] [Fritz!Box] Getting thermostat 117950212336 target heating state
[12/10/2017, 2:10:58 PM] [Fritz!Box] Getting thermostat 117950212336 temperature
[12/10/2017, 2:10:58 PM] [Fritz!Box] Getting thermostat 117950212336 heating state
[12/10/2017, 2:10:58 PM] [Fritz!Box] Getting thermostat 117950212336 battery status
[12/10/2017, 2:10:58 PM] [Fritz!Box] Getting thermostat 117950350088 battery status
[12/10/2017, 2:10:58 PM] [Fritz!Box] Getting thermostat 117950350088 target heating state
[12/10/2017, 2:10:58 PM] [Fritz!Box] Getting thermostat 117950350088 heating state
[12/10/2017, 2:10:58 PM] [Fritz!Box] Getting thermostat 117950350088 target temperature
[12/10/2017, 2:10:58 PM] [Fritz!Box] Getting thermostat 117950350088 temperature
-- home app finished at 2:11:28 PM --
-- home app start from background --
[12/10/2017, 2:14:54 PM] [Fritz!Box] Updating outlet 087610274866
[12/10/2017, 2:14:54 PM] [Fritz!Box] Getting outlet 087610274866 state
[12/10/2017, 2:14:54 PM] [Fritz!Box] Getting outlet 087610274866 temperature
[12/10/2017, 2:14:54 PM] [Fritz!Box] Updating outlet 087610332820
[12/10/2017, 2:14:54 PM] [Fritz!Box] Getting outlet 087610332820 state
[12/10/2017, 2:14:54 PM] [Fritz!Box] Getting outlet 087610332820 temperature
[12/10/2017, 2:14:54 PM] [Fritz!Box] Updating outlet 116570017928
[12/10/2017, 2:14:54 PM] [Fritz!Box] Getting outlet 116570017928 state
[12/10/2017, 2:14:54 PM] [Fritz!Box] Getting outlet 116570017928 temperature
[12/10/2017, 2:14:54 PM] [Fritz!Box] Updating thermostat 117950350088
[12/10/2017, 2:14:54 PM] [Fritz!Box] Getting thermostat 117950350088 temperature
[12/10/2017, 2:14:54 PM] [Fritz!Box] Getting thermostat 117950350088 target temperature
[12/10/2017, 2:14:54 PM] [Fritz!Box] Getting thermostat 117950350088 battery level
[12/10/2017, 2:14:54 PM] [Fritz!Box] Updating thermostat 117950212336
[12/10/2017, 2:14:54 PM] [Fritz!Box] Getting thermostat 117950212336 temperature
[12/10/2017, 2:14:54 PM] [Fritz!Box] Getting thermostat 117950212336 target temperature
[12/10/2017, 2:14:54 PM] [Fritz!Box] Getting thermostat 117950212336 battery level
[12/10/2017, 2:14:54 PM] [Fritz!Box] Updating thermostat 117950210296
[12/10/2017, 2:14:54 PM] [Fritz!Box] Getting thermostat 117950210296 temperature
[12/10/2017, 2:14:54 PM] [Fritz!Box] Getting thermostat 117950210296 target temperature
[12/10/2017, 2:14:54 PM] [Fritz!Box] Getting thermostat 117950210296 battery level
[12/10/2017, 2:14:54 PM] [Fritz!Box] Updating thermostat 117950252408
[12/10/2017, 2:14:54 PM] [Fritz!Box] Getting thermostat 117950252408 temperature
[12/10/2017, 2:14:54 PM] [Fritz!Box] Getting thermostat 117950252408 target temperature
[12/10/2017, 2:14:54 PM] [Fritz!Box] Getting thermostat 117950252408 battery level
[12/10/2017, 2:14:54 PM] [Fritz!Box] Updating thermostat 109711038560
[12/10/2017, 2:14:54 PM] [Fritz!Box] Getting thermostat 109711038560 temperature
[12/10/2017, 2:14:54 PM] [Fritz!Box] Getting thermostat 109711038560 target temperature
[12/10/2017, 2:14:54 PM] [Fritz!Box] Getting thermostat 109711038560 battery level
-- home app update finished at 2:15:14 PM --
````

**Comment by andig** [comment]
@jngmrks Could you please also enable the `debug` option in the platform config? I'd like to see how low the fritz calls actually take. From the log above it's not obvious where the delay comes from. Is `fritz` the only plugin you're having this problem with?

**Comment by andig** [comment]
@xkonni not sure what to make of your normal log file. With the homebridge start at 13:30:07 it takes until 13:30:15 to fetch all characteristics. That seems fine to me. Where exactly in the logfile vs. the actual app handling do you see the problem? How does the problem with the app correlate to the logfile?

**Comment by jngmrks** [comment]
[12/10/2017, 2:30:27 PM] [Fritz!Box] Getting thermostat 109711038560 target heating state
[12/10/2017, 2:30:27 PM] [Fritz!Box] Getting thermostat 109711038560 temperature
[12/10/2017, 2:30:27 PM] [Fritz!Box] Getting thermostat 109711038560 target temperature
[12/10/2017, 2:30:27 PM] [Fritz!Box] Getting thermostat 109711038560 heating state
[12/10/2017, 2:30:27 PM] [Fritz!Box] Getting thermostat 109711038560 battery status
[12/10/2017, 2:30:27 PM] [Fritz!Box] Getting guest WLAN state
[12/10/2017, 2:30:27 PM] [Fritz!Box] Getting thermostat 117950252408 target temperature
[12/10/2017, 2:30:27 PM] [Fritz!Box] Getting thermostat 117950252408 target heating state
[12/10/2017, 2:30:27 PM] [Fritz!Box] Getting thermostat 117950252408 temperature
[12/10/2017, 2:30:27 PM] [Fritz!Box] Getting thermostat 117950252408 heating state
[12/10/2017, 2:30:27 PM] [Fritz!Box] Getting thermostat 117950252408 battery status
[12/10/2017, 2:30:27 PM] [Fritz!Box] Getting thermostat 117950210296 target heating state
[12/10/2017, 2:30:27 PM] [Fritz!Box] Getting thermostat 117950210296 target temperature
[12/10/2017, 2:30:27 PM] [Fritz!Box] Getting thermostat 117950210296 temperature
[12/10/2017, 2:30:27 PM] [Fritz!Box] Getting thermostat 117950210296 heating state
[12/10/2017, 2:30:27 PM] [Fritz!Box] Getting thermostat 117950210296 battery status
[12/10/2017, 2:30:27 PM] [Fritz!Box] Getting outlet 116570017928 state
[12/10/2017, 2:30:27 PM] [Fritz!Box] Getting outlet 116570017928 temperature
[12/10/2017, 2:30:27 PM] [Fritz!Box] Getting thermostat 117950212336 battery status
[12/10/2017, 2:30:27 PM] [Fritz!Box] Getting thermostat 117950212336 target heating state
[12/10/2017, 2:30:27 PM] [Fritz!Box] Getting thermostat 117950212336 temperature
[12/10/2017, 2:30:27 PM] [Fritz!Box] Getting thermostat 117950212336 target temperature
[12/10/2017, 2:30:27 PM] [Fritz!Box] Getting thermostat 117950212336 heating state
[12/10/2017, 2:30:27 PM] [Fritz!Box] Getting thermostat 117950350088 battery status
[12/10/2017, 2:30:27 PM] [Fritz!Box] Getting thermostat 117950350088 target heating state
[12/10/2017, 2:30:27 PM] [Fritz!Box] Getting thermostat 117950350088 heating state
[12/10/2017, 2:30:27 PM] [Fritz!Box] Getting thermostat 117950350088 target temperature
[12/10/2017, 2:30:27 PM] [Fritz!Box] Getting thermostat 117950350088 temperature
[12/10/2017, 2:30:27 PM] [Fritz!Box] > getTempTarget ("e3613f9771821405","109711038560")
[12/10/2017, 2:30:28 PM] [Fritz!Box] < getTempTarget 22
[12/10/2017, 2:30:28 PM] [Fritz!Box] > getTemperature ("e3613f9771821405","109711038560")
[12/10/2017, 2:30:28 PM] [Fritz!Box] < getTemperature 22.5
[12/10/2017, 2:30:28 PM] [Fritz!Box] > getTempTarget ("e3613f9771821405","109711038560")
[12/10/2017, 2:30:29 PM] [Fritz!Box] < getTempTarget 22
[12/10/2017, 2:30:29 PM] [Fritz!Box] > getTempTarget ("e3613f9771821405","109711038560")
[12/10/2017, 2:30:29 PM] [Fritz!Box] < getTempTarget 22
[12/10/2017, 2:30:29 PM] [Fritz!Box] > getBatteryCharge ("e3613f9771821405","109711038560")
[12/10/2017, 2:30:32 PM] [Fritz!Box] < getBatteryCharge 20
[12/10/2017, 2:30:32 PM] [Fritz!Box] > getGuestWlan ("e3613f9771821405")
[12/10/2017, 2:30:33 PM] [Fritz!Box] < getGuestWlan {"autoupdate":true,"activate_guest_access":false,"guest_ssid":"xxx","wpa_key":"xxx","push_service":true,"group_access":true,"untrusted":false,"showlawtext":false,"portal_redirect_enabled":false,"portal_redirect_url":"","user_isolation":false,"down_time_activ":true,"disconnect_guest_access":true,"sec_mode":"3","down_time_value":"120"}
[12/10/2017, 2:30:33 PM] [Fritz!Box] > getTempTarget ("e3613f9771821405","117950252408")
[12/10/2017, 2:30:34 PM] [Fritz!Box] < getTempTarget 22
[12/10/2017, 2:30:34 PM] [Fritz!Box] > getTempTarget ("e3613f9771821405","117950252408")
[12/10/2017, 2:30:34 PM] [Fritz!Box] < getTempTarget 22
[12/10/2017, 2:30:34 PM] [Fritz!Box] > getTemperature ("e3613f9771821405","117950252408")
[12/10/2017, 2:30:34 PM] [Fritz!Box] < getTemperature 22.5
[12/10/2017, 2:30:34 PM] [Fritz!Box] > getTempTarget ("e3613f9771821405","117950252408")
[12/10/2017, 2:30:35 PM] [Fritz!Box] < getTempTarget 22
[12/10/2017, 2:30:35 PM] [Fritz!Box] > getBatteryCharge ("e3613f9771821405","117950252408")
[12/10/2017, 2:30:37 PM] [Fritz!Box] < getBatteryCharge 20
[12/10/2017, 2:30:37 PM] [Fritz!Box] > getTempTarget ("e3613f9771821405","117950210296")
[12/10/2017, 2:30:37 PM] [Fritz!Box] < getTempTarget 22
[12/10/2017, 2:30:37 PM] [Fritz!Box] > getTempTarget ("e3613f9771821405","117950210296")
[12/10/2017, 2:30:37 PM] [Fritz!Box] < getTempTarget 22
[12/10/2017, 2:30:37 PM] [Fritz!Box] > getTemperature ("e3613f9771821405","117950210296")
[12/10/2017, 2:30:38 PM] [Fritz!Box] < getTemperature 22.5
[12/10/2017, 2:30:38 PM] [Fritz!Box] > getTempTarget ("e3613f9771821405","117950210296")
[12/10/2017, 2:30:38 PM] [Fritz!Box] < getTempTarget 22
[12/10/2017, 2:30:38 PM] [Fritz!Box] > getBatteryCharge ("e3613f9771821405","117950210296")
[12/10/2017, 2:30:40 PM] [Fritz!Box] < getBatteryCharge 20
[12/10/2017, 2:30:40 PM] [Fritz!Box] > getSwitchState ("e3613f9771821405","116570017928")
[12/10/2017, 2:30:41 PM] [Fritz!Box] < getSwitchState false
[12/10/2017, 2:30:41 PM] [Fritz!Box] > getTemperature ("e3613f9771821405","116570017928")
[12/10/2017, 2:30:41 PM] [Fritz!Box] < getTemperature -1.5
[12/10/2017, 2:30:41 PM] [Fritz!Box] > getBatteryCharge ("e3613f9771821405","117950212336")
[12/10/2017, 2:30:43 PM] [Fritz!Box] < getBatteryCharge 20
[12/10/2017, 2:30:43 PM] [Fritz!Box] > getTempTarget ("e3613f9771821405","117950212336")
[12/10/2017, 2:30:43 PM] [Fritz!Box] < getTempTarget 22
[12/10/2017, 2:30:43 PM] [Fritz!Box] > getTemperature ("e3613f9771821405","117950212336")
[12/10/2017, 2:30:43 PM] [Fritz!Box] < getTemperature 22
[12/10/2017, 2:30:43 PM] [Fritz!Box] > getTempTarget ("e3613f9771821405","117950212336")
[12/10/2017, 2:30:44 PM] [Fritz!Box] < getTempTarget 22
[12/10/2017, 2:30:44 PM] [Fritz!Box] > getTempTarget ("e3613f9771821405","117950212336")
[12/10/2017, 2:30:44 PM] [Fritz!Box] < getTempTarget 22
[12/10/2017, 2:30:44 PM] [Fritz!Box] > getBatteryCharge ("e3613f9771821405","117950350088")
[12/10/2017, 2:30:46 PM] [Fritz!Box] < getBatteryCharge 10
[12/10/2017, 2:30:46 PM] [Fritz!Box] > getTempTarget ("e3613f9771821405","117950350088")
[12/10/2017, 2:30:46 PM] [Fritz!Box] < getTempTarget 22
[12/10/2017, 2:30:46 PM] [Fritz!Box] > getTempTarget ("e3613f9771821405","117950350088")
[12/10/2017, 2:30:47 PM] [Fritz!Box] < getTempTarget 22
[12/10/2017, 2:30:47 PM] [Fritz!Box] > getTempTarget ("e3613f9771821405","117950350088")
[12/10/2017, 2:30:47 PM] [Fritz!Box] < getTempTarget 22
[12/10/2017, 2:30:47 PM] [Fritz!Box] > getTemperature ("e3613f9771821405","117950350088")
[12/10/2017, 2:30:48 PM] [Fritz!Box] < getTemperature 23
-- now 21 seconds -
only have the fritz and camera-ffmpeg plugin running (last one disabled for test)

**Comment by xkonni** [comment]
The time is perfectly fine.

But it should not occur when opening e.g. the quick toggles. That could deliver cached values instead. Or cached first, trigger update, send updated values.

The entire toggles are unusable until the update completes.

Thanks for your help!

**Comment by jngmrks** [comment]
Tried now with the #automode, same time needed. Must wait every time about 20-30 seconds before i could see or change anything. Seems there is no cache. Debug shows me an update on each request. The home app is only working if all data are complete.

Siri-Request:

[12/10/2017, 2:57:42 PM] [Fritz!Box] Getting thermostat 117950350088 battery level
[12/10/2017, 2:57:42 PM] [Fritz!Box] > getBatteryCharge ("459c8e250d644f8f","117950350088")
[12/10/2017, 2:57:44 PM] [Fritz!Box] < getBatteryCharge 10

**Comment by andig** [comment]
Ok. Let me try to summarise: whenever the home app is opened or reopened, it takes 20-30 seconds before the ui becomes responsive because it is apparently updating all the switches and sensors?

This does indeed not make too much sense since they are already updated regularly in the background. I'll check if returning cached characteristics can fix this.

Do you have any example of a plugin with a sizeable amount of devices that does behave properly?

**Comment by xkonni** [comment]
@jngmrks that i wanted to make sure that it's caused by this plugin and not some other.
@andig that sounds perfect! unfortunately the other plugins i'm using don't need to collect their state in the background.

**Comment by jngmrks** [comment]
The collecting of states in the background is for updates to the home app. See https://github.com/andig/homebridge-fritz/issues/14. But its anyway not working.

**Comment by andig** [comment]
I've pushed a fix for this, install using 

    npm install --unsafe-perm git://github.com/andig/homebridge-fritz#update

@jngmrks #14 has meanwhile been fixed, too.

**Comment by andig** [comment]
The only drawback right now is that immediately after starting homebridge all accessories will immediately show falsey state as the devices have not yet been physically updated. I won't change that though as I consider the impact low.

I would appreciate a small paypal donation if you're happy with the result of the ticket resolution.

**Comment by jngmrks** [comment]
Wow, the loading time is now almost zero. The false state after homebridge start only takes 1-2 seconds per device. Now also devices are updated as single device and not all at once. Good Job!

As is can see #14 was the fix for #30.

**Comment by xkonni** [comment]
That's awesome, updated and works perfectly.

I agree, the < 60s after startup don't matter. 

Thank you so much for this!

**Comment by andig** [comment]
Thank you for helping me to finally understand how homebridge/kit is operating under the hood. No more black box :)

Released as `0.5.12`
